### PR TITLE
Round should avoid clearing s, save a string

### PR DESCRIPTION
**Pull request by smarterclayton**
Instead of saving bytes, save a string, which makes String() faster
and does not unduly penalize marshal. During parse, save the string
if it is in canonical form.

@wojtek-t @lavalamp this makes quantity.String() faster for a few cases
where it matters. We were also not clearing s properly before on Round()

### CODE REVIEW

**Comment by smarterclayton** [comment]
@k8s-bot test this issue #IGNORE (to pick up the latest changes this morning)

**Comment by smarterclayton** [comment]
@wojtek-t if you can review this I'd appreciate it.  It's a correctness issue w.r.t unmarshalling so I'm not positive it's not causing issues.

**Comment by wojtek-t** [comment]
@smarterclayton - I'm not very comfortable reviewing this one. @lavalamp - can you please take a look?

File: pkg/api/resource/quantity.go [hunk 1]
```diff
@@ -502,7 +519,7 @@ func (q *Quantity) RoundUp(scale Scale) bool {
 // Add adds the provide y quantity to the current value. If the current value is zero,
 // the format of the quantity will be updated to the format of y.
 func (q *Quantity) Add(y Quantity) {
-	q.s = nil
+	q.s = ""
```
Comment (lavalamp) [inline pkg/api/resource/quantity.go]: I'm scared someone is going to omit this line some place where it's needed...

File: pkg/api/resource/quantity_test.go [hunk 2]
```diff
@@ -674,6 +674,13 @@ func TestQuantityString(t *testing.T) {
 		if e, a := item.expect, got; e != a {
 			t.Errorf("%#v: expected %v, got %v", item.in, e, a)
 		}
+		q, err := ParseQuantity(item.expect)
+		if err != nil {
+			t.Errorf("%#v: unexpected error: %v", item.expect, err)
+		}
+		if len(q.s) == 0 || q.s != item.expect {
+			t.Errorf("%#v: did not copy canonical string on parse: %s", item.expect, q.s)
```
Comment (lavalamp) [inline pkg/api/resource/quantity_test.go]: Test the converse? (that it doesn't cache noncanonical strings?)

**Comment by lavalamp** [comment]
LGTM one request for another test

File: pkg/api/resource/quantity.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/api/resource/quantity.go]: Yes.  We may need a more holistic quantity stress tester (I did one
for add and remove, but we might want to get more formal).

Comment (smarterclayton) [inline pkg/api/resource/quantity.go]: Added a test, also added one more guard for non-canonical strings (a
leading 0 in the shifted numerator+denominator cannot be canonical).
The parse routine always returns 0 for numerator (even when it has a
leading .) and so the left most position being 0 means the value
cannot be canonical.

**Comment by smarterclayton** [comment]
Test added, adding label.

**Comment by smarterclayton** [comment]
@k8s-bot test this issue #20708

**Comment by smarterclayton** [comment]
xref https://github.com/kubernetes/kubernetes/issues/20576 instead

**Pull request merged**